spy/facility.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package spy
+
+import (
+	"sync"
+
+	"go.uber.org/zap"
+)
+
+// A Log is an encoding-agnostic representation of a log message.
+type Log struct {
+	Level  zap.Level
+	Msg    string
+	Fields []zap.Field
+}
+
+// A Sink stores Log structs.
+type Sink struct {
+	sync.Mutex
+
+	logs []Log
+}
+
+// WriteLog writes a log message to the LogSink.
+func (s *Sink) WriteLog(lvl zap.Level, msg string, fields []zap.Field) {
+	s.Lock()
+	log := Log{
+		Msg:    msg,
+		Level:  lvl,
+		Fields: fields,
+	}
+	s.logs = append(s.logs, log)
+	s.Unlock()
+}
+
+// Logs returns a copy of the sink's accumulated logs.
+func (s *Sink) Logs() []Log {
+	var logs []Log
+	s.Lock()
+	logs = append(logs, s.logs...)
+	s.Unlock()
+	return logs
+}
+
+// Facility implements a zap.Facility that captures Log records.
+type Facility struct {
+	sync.Mutex
+	zap.LevelEnabler
+	sink    *Sink
+	context []zap.Field
+}
+
+// With creates a sub spy facility so that all log records recorded under it
+// have the given fields attached.
+func (sf *Facility) With(fields ...zap.Field) zap.Facility {
+	return &Facility{
+		LevelEnabler: sf.LevelEnabler,
+		sink:         sf.sink,
+		context:      append(sf.context, fields...),
```

Comment:
This is dangerous and can allow With to overwrite fields from another
facility.
For example, consider:

	f, sink := spy.New(zap.DebugLevel)
	l1 := zap.New(f).With(zap.Int("l1 both", 1)).With(zap.Int("l1 both 2", 1))
	l2 := l1.With(zap.Int("l2 both", 2))
	l3 := l2.With(zap.Int("l3 only", 3))
	l4 := l2.With(zap.Int("l4 only", 4))
	l3.Info("l3")
	l4.Info("l4")

The l3.Info log will print the "l4 only" field because the context
has been overwritten by the append to the shared slice.

I think you probably want:

      context:      append(sf.context[0:len(sf.context):len(sf.context)], fields...),

to force the whole slice to be copied each time.